### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by kaaholst**
Implement "Play from here"

### CODE REVIEW

File: Squeezer/src/main/java/uk/org/ngo/squeezer/itemlist/AlarmsActivity.java [hunk 1]
```diff
@@ -181,20 +181,20 @@ protected void orderPage(@NonNull ISqueezeService service, int start) {
...
+                    }
 
-            mAlarmPlaylists.addAll(items);
-            if (start + items.size() >= count) {
-                getUIThreadHandler().post(new Runnable() {
-                    @Override
-                    public void run() {
+                    mAlarmPlaylists.addAll(items);
+                    if (start + items.size() >= count) {
                         mAlarmView.setAlarmPlaylists(mAlarmPlaylists);
```
Comment (nikclayton) [inline Squeezer/src/main/java/uk/org/ngo/squeezer/itemlist/AlarmsActivity.java]: I'll merge as is for the moment, but what if just this line was changed to be:

```
mAlarmView.setAlarmPlaylists(ImmutableList.copyOf(mAlarmPlaylists);
```
instead?

Comment (kaaholst) [inline Squeezer/src/main/java/uk/org/ngo/squeezer/itemlist/AlarmsActivity.java]: That seems like a good idea. 

Note that the semantics is slightly different, as the UI is not updated until we have received all the items, which also is how it worked originally.

One drawback is that copy of the list is thrown away when AlarmView is done with it, even more so as all AlarmView does with it, is add a few category items to it.

Comment (kaaholst) [inline Squeezer/src/main/java/uk/org/ngo/squeezer/itemlist/AlarmsActivity.java]: Actually the semantics are the same the UI isn't updated until all items are received.

File: Squeezer/build.gradle [hunk 2]
```diff
@@ -39,7 +39,7 @@ dependencies {
     compile 'com.google.code.findbugs:jsr305:2.0.2'
 
     // EventBus, https://github.com/greenrobot/EventBus.
-    compile project(':libs:eventbus:EventBus')
+    compile 'de.greenrobot:eventbus:2.4.1'
```
Comment (nikclayton) [inline Squeezer/build.gradle]: I don't know that 2.4.1 actually fixes this.

Ahead of releasing 1.4.1 I've been launching Squeezer on some emulators running API 7, and they crash with

05-15 18:27:25.134 3064-3064/uk.org.ngo.squeezer W/System.err: java.lang.NoSuchMethodException
05-15 18:27:25.144 3064-3064/uk.org.ngo.squeezer W/System.err:     at java.lang.Class.getDeclaredMethods(Native Method)
05-15 18:27:25.154 3064-3064/uk.org.ngo.squeezer W/System.err:     at java.lang.ClassCache.getDeclaredMethods(ClassCache.java:153)
05-15 18:27:25.154 3064-3064/uk.org.ngo.squeezer W/System.err:     at java.lang.Class.getDeclaredMethods(Class.java:792)
05-15 18:27:25.154 3064-3064/uk.org.ngo.squeezer W/System.err:     at a.b.a.o.a(SubscriberMethodFinder.java:76)
05-15 18:27:25.165 3064-3064/uk.org.ngo.squeezer W/System.err:     at a.b.a.c.register(EventBus.java:164)
05-15 18:27:25.165 3064-3064/uk.org.ngo.squeezer W/System.err:     at a.b.a.c.register(EventBus.java:144)
05-15 18:27:25.165 3064-3064/uk.org.ngo.squeezer W/System.err:     at uk.org.ngo.squeezer.service.SqueezeService$EventBus.register(SqueezeService.java:1691)
05-15 18:27:25.165 3064-3064/uk.org.ngo.squeezer W/System.err:     at uk.org.ngo.squeezer.service.SqueezeService.onCreate(SqueezeService.java:220)
05-15 18:27:25.185 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.app.ActivityThread.handleCreateService(ActivityThread.java:2780)
05-15 18:27:25.185 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.app.ActivityThread.access$3200(ActivityThread.java:119)
05-15 18:27:25.185 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1917)
05-15 18:27:25.185 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:99)
05-15 18:27:25.185 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.os.Looper.loop(Looper.java:123)
05-15 18:27:25.195 3064-3064/uk.org.ngo.squeezer W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:4363)
05-15 18:27:25.195 3064-3064/uk.org.ngo.squeezer W/System.err:     at java.lang.reflect.Method.invokeNative(Native Method)
05-15 18:27:25.195 3064-3064/uk.org.ngo.squeezer W/System.err:     at java.lang.reflect.Method.invoke(Method.java:521)
05-15 18:27:25.195 3064-3064/uk.org.ngo.squeezer W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:860)
05-15 18:27:25.205 3064-3064/uk.org.ngo.squeezer W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:618)
05-15 18:27:25.205 3064-3064/uk.org.ngo.squeezer W/System.err:     at dalvik.system.NativeStart.main(Native Method)

**Comment by kaaholst** [comment]
I think it should fix it.

EventBus [issue 149](https://github.com/greenrobot/EventBus/issues/149#issuecomment-76316437) references a [commit](https://github.com/greenrobot/EventBus/commit/be04a2de66981e819f40afd389e417e98c94c7da), which implement the recommended workaround by [@yuzeh](https://github.com/yuzeh), who's fork we previously referenced in Squeezer.

I haven't been able to trigger the problem though, so I don't know if 2.4.1 actually fixes it.

I don't understand the supplied stacktrace though. When I open SubscriberMethodFinder.java from eventbus-2.4.1-sources.jar, line 76 is: `Method[] methods = clazz.getDeclaredMethods();`, and this line is protected by a try-catch (Throwable) and falls back to getMethods(). So how can it throw a NoSuchMethodException?

File: Squeezer/build.gradle (hunk 2 above)
Comment (kaaholst) [inline Squeezer/build.gradle]: Just did a few more tests based on the [faq ](http://greenrobot.org/eventbus/documentation/faq/)mentioned in [issue 149](https://github.com/greenrobot/EventBus/issues/149) . 

> The exception is thrown if the class has methods with a parameter that is unavailable to the API level of the device.

When I launch squeezer on an API 7 (Android 2.1) emulator I get a crash on line 81 of SubscriberMethodFinder.java, which makes sense as this is the fallback ("plan b") applied by EventBus when getDeclaredMethods fails.

I then comment out the `Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP` section of updateOngoingNotification and the crash disappears.

I suspect (but I'm guessing here as I haven't done the test) that the crash would happen also with the EventBus fork pulled in and on EventBus 3.0 (unless we use an index).

**Comment by nikclayton** [comment]
I've got a nagging feeling this was caused by the Instant Run feature in newer versions of Android Studio (where it hot-patches new code in to the app instead of deploying a fresh APK). I turned that off, and can't reproduce.

1.4.1-beta contains 2.4.1 now (no crashes in Crashlytics that I can see), and the eventbus3 branch has https://github.com/nikclayton/android-squeezer/commit/15e08b029e8af97516152de71acc55cba6260d83 which is all the work (I think) necessary to use eventbus v3. I'll merge that in after 1.4.1 is released.

**Pull request merged**